app/code/Magento/Store/Model/Store.php

```diff
@@ -463,10 +463,10 @@ protected function _getValidationRulesBeforeSave()
         $storeLabelRule->setMessage(__('Name is required'), \Zend_Validate_NotEmpty::IS_EMPTY);
         $validator->addRule($storeLabelRule, 'name');
 
-        $storeCodeRule = new \Zend_Validate_Regex('/^[a-z]+[a-z0-9_]*$/');
+        $storeCodeRule = new \Zend_Validate_Regex('/^[a-zA-Z]+[a-z0-9_]*$/');
```

Comment:
This expression will not allow code with uppercase letter after number or underscore (e.g. `a1A`). Instead of modifying regexp it's better to apply modifier `i` ([PCRE_CASELESS](http://php.net/manual/en/reference.pcre.pattern.modifiers.php)) to make case-insensitive comparison.

Reply:
Done

Reply:
I would suggest you change the regex to be 

 $storeCodeRule = new \Zend_Validate_Regex('/^[A-z]+[a-z0-9_]*$/');

[A-z] is the same as [a-zA-Z] but shorter and more concise